File: +ndi/+fun/timestamp.m
function timestamp_string = timestamp()
% TIMESTAMP - return a current time stamp string
%
% TIMESTAMP_STRING = ndi.fun.timestamp()
%
% Returns a current time stamp string using the expression:
%   TIMESTAMP_STRING = char(datetime('now','TimeZone','UTCLeapSeconds'))
%
% The string is checked to make sure that the seconds are not
% "60.000", which can occur due to rounding and which can cause a 
% validation error when the data is included in a database.
% In that case, the seconds are set to "59.999".
%
% Example:
%   ts_st = ndi.fun.timestamp()
%

timestamp_string = char(datetime('now','TimeZone','UTCLeapSeconds'));

if strcmp(timestamp_string(end-6:end-1),'60.000'),
	timestamp_string(end-6:end-1) = '59.999';
end;
 

File: ndi-python/ndi/fun/timestamp.py
"""
Generate UTC timestamps for NDI.

This module provides timestamp generation functionality for NDI documents
and logging.

Ported from MATLAB: src/ndi/+ndi/+fun/timestamp.m
"""

from datetime import datetime, timezone


def timestamp() -> str:
    """
    Return a current UTC timestamp string.

    Returns a current timestamp string in UTC with leap seconds handling.
    The format is ISO 8601 compatible. The string is checked to ensure
    seconds are not "60.000" (which can occur due to rounding and cause
    database validation errors). In that case, seconds are set to "59.999".

    Returns:
        str: UTC timestamp string in ISO format

    Example:
        >>> ts = timestamp()
        >>> print(ts)
        '2025-11-16T12:34:56.789000+00:00'

    Notes:
        - Uses UTC timezone
        - Handles leap second rounding (60.000 -> 59.999)
        - Compatible with database storage
        - ISO 8601 format
    """
    # Get current UTC time
    timestamp_string = datetime.now(timezone.utc).isoformat()

    # Check for and fix the 60.000 seconds issue
    # This can happen due to leap seconds and rounding
    if '60.000' in timestamp_string:
        timestamp_string = timestamp_string.replace('60.000', '59.999')

    return timestamp_string


def timestamp_matlab_format() -> str:
    """
    Return timestamp in MATLAB-compatible format.

    Returns:
        str: Timestamp string matching MATLAB's datetime format

    Example:
        >>> ts = timestamp_matlab_format()
        >>> print(ts)
        '16-Nov-2025 12:34:56'
    """
    # MATLAB format: dd-MMM-yyyy HH:mm:ss
    dt = datetime.now(timezone.utc)
    return dt.strftime('%d-%b-%Y %H:%M:%S')
